### PR TITLE
Prepare for TypeScript null reference checking

### PR DESCRIPTION
**Pull request by lindner**
- Add types in a lot more places.
- Reformat random.ts from 4 to 2 space indent
- Add missing imports
- Fix variable aliasing (reusing as different types)
- Activate strictFunctionTypes

### CODE REVIEW

File: runtime/ts/storage/crdt-collection-model.ts [hunk 1]
```diff
@@ -85,15 +90,15 @@ export class CrdtCollectionModel {
   }
 
   // [{id, value, keys: []}]
-  toLiteral(): {id, value, keys}[] {
-    const result: {id, value, keys}[] = [];
+  toLiteral(): Model[] {
```
Comment (shans) [inline runtime/ts/storage/crdt-collection-model.ts]: This change makes me realize that 'Model' isn't a good name for the interface. We don't want to have a `CrdtCollectionModel`'s `Model`, that's confusing. Even more confusing is that this interface represents a single `Entity`, not all of the entities in the model.

How about `Entity` or `EntityData`?

File: runtime/ts/storage/storage-provider-base.ts [hunk 2]
```diff
@@ -25,14 +25,14 @@ export class StorageProviderBase {
...
+  name: string|undefined;
   source: {}|null;
   description: string;
 
-  constructor(type, name, id, key) {
+  constructor(type: Type, name: string|undefined, id: string, key: string) {
     assert(id, 'id must be provided when constructing StorageProviders');
     assert(!type.hasUnresolvedVariable, 'Storage types must be concrete');
-    const trace = Tracing.start({cat: 'handle', name: 'StorageProviderBase::constructor', args: {type: type.key, name}});
+    const trace = Tracing.start({cat: 'handle', name: 'StorageProviderBase::constructor', args: {type, name}});
```
[the remark this answers is not in the document]
Comment (lindner) [reply runtime/ts/storage/storage-provider-base.ts]: done

File: runtime/ts/storage/storage-provider-factory.ts [hunk 3]
```diff
@@ -8,7 +8,10 @@
 
 import {InMemoryStorage} from './in-memory-storage';
 import {FirebaseStorage} from './firebase-storage';
```
[the remark this answers is not in the document]
Comment (lindner) [reply runtime/ts/storage/storage-provider-factory.ts]: assume you mean add .js suffixes..

File: runtime/ts/storage/storage-provider-factory.ts [hunk 4]
```diff
@@ -17,28 +20,28 @@ export class StorageProviderFactory {
     this._storageInstances = {'in-memory': new InMemoryStorage(arcId), 'firebase': new FirebaseStorage(arcId)};
   }
 
-  _storageForKey(key) {
+  private _storageForKey(key: string): InMemoryStorage | FirebaseStorage {
```
[the remark this answers is not in the document]
Comment (lindner) [reply runtime/ts/storage/storage-provider-factory.ts]: done

File: runtime/ts/storage/storage-provider-factory.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (shans) [reply runtime/ts/storage/storage-provider-factory.ts]: yup, thanks

File: runtime/ts/type.ts [hunk 5]
```diff
@@ -71,39 +71,39 @@ export class Type {
     this.data = data;
   }
 
-  static newEntity(entity : Schema) {
+  static newEntity(entity : Schema): Type {
```
[the remark this answers is not in the document]
Comment (lindner) [reply runtime/ts/type.ts]: Removed.

FWIW here's the internal guide on inference:

http://go/tsstyle#type-inference

I do think that annotating return types makes some sense, especially for newVariable() which returns a type (not a Variable...)

File: runtime/ts/storage/storage-provider-factory.ts [hunk 6]
```diff
@@ -17,28 +20,28 @@ export class StorageProviderFactory {
...
     const protocol = key.split(':')[0];
     return this._storageInstances[protocol];
   }
 
-  async share(id, type, key) {
+  async share(id: string, type: Type, key: string): Promise<StorageProviderBase|null> {
```
[the remark this answers is not in the document]
Comment (lindner) [reply runtime/ts/storage/storage-provider-factory.ts]: This is necessary to designate that the return is nullable.  It also forces all subclasses to implement the same type signature for this method.

This explicit contract improves the documentation of this base class.  When I was writing pouch code I had no idea what each of these methods were for, and exactly what they should return. 

Also until we outlaw `any` then it's very difficult to know if/when typescript correctly infers the types.  There's also http://go/tsstyle#any-type to consider....

[the remark this answers is not in the document]
Comment (shans) [reply runtime/ts/storage/storage-provider-factory.ts]: This CL adds return types for a bunch of methods, not just the nullable ones. I'm OK with us specifying return types when the inference is insufficient; this should generally be indicated by the compiler *before* we proceed with adding a stronger type.

Documentation of contracts is also a reasonable goal, but again this applies only to a fraction of the return type annotations you've added within this CL.

The approach that the google3 style guide authors have taken is to make it as inconvenient as possible to make use of the any type. This makes sense when evaluating typeScript as a replacement for closure; however the approach is a pretty bad impedance mismatch for non-closure JS and IMHO is not a good direction for us. Instead, forcing `any` to be explicitly annotated should be enough for us to signal that we're aware we are moving away from type protection when warranted.

Until we outlaw *implicit* `any`, sure, we don't know if the inference is correct. However we are no worse off than we currently are in that regards; so it's completely acceptable to wait until we turn that rule on in the compiler before adding annotations in the places that need them.

File: runtime/ts/type.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (shans) [reply runtime/ts/type.ts]: Honestly if that's a point of confusion we'd be better to rename the function...

**Comment by lindner** [comment]
I think that this has exposed a larger philosophical discussion.  I've grabbed some time with the Firebase JS SDK TL and TypeScript Skeptic to discuss what they consider a sane environment.

**Pull request closed without merging**